daemon/containerd: Extract `createOrReplaceImage`

vvoland opened the pull request:
- fixes: https://github.com/moby/moby/issues/48591
- addresses: (`build` and `imports`, but not `load`); https://github.com/moby/moby/issues/48907


**- What I did**

Extract a method that creates an image if it doesn't exists and replaces it otherwise. Use it in other places which also tried to do the same thing.


**- How to verify it**
Tests should pass

```markdown changelog
containerd image store: Fix `commit`, `import` and `build` not preserving replaced image as a dangling.
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -42,8 +50,8 @@ func (i *ImageService) TagImage(ctx context.Context, imageID image.ID, newTag re
-		if replacedImg.Target.Digest == targetImage.Target.Digest {
-			i.LogImageEvent(imageID.String(), reference.FamiliarString(newTag), events.ActionTag)
+		if replacedImg.Target.Digest == newImg.Target.Digest {
+			i.LogImageEvent(newImg.Name, imageFamiliarName(newImg), events.ActionTag)
```
This is changing `id` to `name`, correct? Mostly considering here that name could be easier to read, but `ID` may work also in case where there's no name associated. I recall we had some similar discussion around "network" events (I think) at some point, where we could not add a name to all of the related events, which made it more difficult to correlate events.

vvoland replied:
Ah yes, good catch, this should be the ID

vvoland commented on daemon/containerd/image_builder.go:
```diff
@@ -524,18 +524,11 @@ func (i *ImageService) createImageOCI(ctx context.Context, imgToCreate imagespec
 		img.Labels[imageLabelClassicBuilderFromScratch] = "1"
 	}
 
-	createdImage, err := i.images.Update(ctx, img)
```
This was calling `Update` specifically to handle the case where the image already exists - if update fails due to non found error, it would do the `Create` instead.

The new function handles the case a bit differently - instead of just updating the image, it will first remove it if it exists (or rename to a dangling reference if it's the last tag to that image).

vvoland committed:
c8d/tag: Extract createOrReplaceImage

Extract a method that creates an image if it doesn't exists and replaces
it otherwise.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/commit,import,build: Preserve overriden image

If the commit, import or build targets an image tag that already exists,
persist the old image under a dangling reference instead of just
deleting it.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -42,8 +56,10 @@ func (i *ImageService) TagImage(ctx context.Context, imageID image.ID, newTag re
-			i.LogImageEvent(imageID.String(), reference.FamiliarString(newTag), events.ActionTag)
+		if replacedImg.Target.Digest == newImg.Target.Digest {
+			if !creatingDangling {
+				i.LogImageEvent(replacedImg.Target.Digest.String(), imageFamiliarName(newImg), events.ActionTag)
```
Note to self; this one  will conflict with a PR I had pending (I need to update that one, but it's less important  than this PR);

- https://github.com/moby/moby/pull/49014

thaJeztah commented on daemon/containerd/image_import.go:
```diff
@@ -145,8 +145,7 @@ func (i *ImageService) ImportImage(ctx context.Context, ref reference.Named, pla
 
-	err = i.saveImage(ctx, img)
-	if err != nil {
+	if err = i.createOrReplaceImage(ctx, img); err != nil {
```
Discussed briefly on slack, and let's keep it as-is for now.

thaJeztah reviewed (approved):
LGTM

@laurazard ptal 🤗

thompson-shaun commented:
Can we close https://github.com/moby/moby/issues/48591 since this got in? @thaJeztah @vvoland
